### PR TITLE
Integrate multihop relayer

### PR DESCRIPTION
**Pull request by 7AC**
Integrates the Go relayer modified to support polymer (see https://github.com/polymerdao/relayer-ibcx/pull/3)

TODO: multiple channel support is coming in the next PR

Closes https://github.com/open-ibc/ibc-sdk/issues/64, closes https://github.com/polymerdao/polymerase/issues/764, closes https://github.com/polymerdao/polymerase/issues/763

### CODE REVIEW

File: src/lib/relayers/ibc.ts [hunk 1]
```diff
@@ -55,74 +111,144 @@ export class IBCRelayer {
...
+      if (obj.clientId.startsWith(tendermintClientPrefix)) {
+        existingSrcConnections.add(obj.id)
+      }
+    })
+    const existingDstConnections = new Set<string>()
+    ;(await path.dst.chainClient.ibcConnections()).connections.forEach((obj) => {
+      if (obj.clientId.startsWith(tendermintClientPrefix)) {
+        existingDstConnections.add(obj.id)
+      }
+    })
```
Comment (nicopernas) [inline src/lib/relayers/ibc.ts:183]: this is calling `path.dst.chainClient.ibcConnections()` twice. maybe you meant to use `src` first and `dst` later?

File: src/lib/relayers/ibc.ts [hunk 2]
```diff
@@ -55,74 +111,144 @@ export class IBCRelayer {
...
-    }
+  async connect(path: Path, pathConfig: any): Promise<any> {
+    const existingSrcConnections = new Set<string>()
+    ;(await path.dst.chainClient.ibcConnections()).connections.forEach((obj) => {
+      if (obj.clientId.startsWith(tendermintClientPrefix)) {
+        existingSrcConnections.add(obj.id)
+      }
+    })
+    const existingDstConnections = new Set<string>()
+    ;(await path.dst.chainClient.ibcConnections()).connections.forEach((obj) => {
```
Comment (nicopernas) [inline src/lib/relayers/ibc.ts:179]: nit: this `;(await ...)` is rather confusing. maybe split the call in two?

File: src/lib/relayers/ibc.ts [hunk 3]
```diff
@@ -55,74 +111,144 @@ export class IBCRelayer {
...
+        existingDstConnections.add(obj.id)
+      }
+    })
+    log.verbose(`creating client for ${pathConfig.name}`)
+    await this.exec([this.binary, 'transact', 'clients', pathConfig.name]).catch((e) => {
+      log.error(e)
+      throw new Error(e)
+    })
     // TODO: this is hacky
     await sleep(10_000)
```
Comment (nicopernas) [inline src/lib/relayers/ibc.ts:190]: we can probably get rid of this now right?

File: src/lib/relayers/ibc.ts [hunk 4]
```diff
@@ -55,74 +111,144 @@ export class IBCRelayer {
...
+    log.verbose(`creating connection for ${pathConfig.name}`)
+    await this.exec([this.binary, 'transact', 'connection', pathConfig.name]).catch((e) => {
+      log.error(e)
+      throw new Error(e)
+    })
+    let newSrcConnectionId: string | undefined
+    let newSrcClientId: string | undefined
+    let newDstConnectionId: string | undefined
+    let newDstClientId: string | undefined
+    await waitUntil(
```
Comment (nicopernas) [inline src/lib/relayers/ibc.ts:201]: nit: this could use some refactoring?

File: src/lib/vibc/channel.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+  // TODO: use grpc
+  const polymerClient = self.cosmos.cliClient.CosmosChainClient.fromRunningContainer(polymer.chain)
+  const dstClient = self.cosmos.cliClient.CosmosChainClient.fromRunningContainer(ibcB.chain)
+
+  // step 1: fetch IBC-side clients/connections
+  const singleHopPath = await ibcPathFromChainClients(polymerClient, dstClient, polymer.chain.Name, ibcB.chain.Name)
+  // TODO: reuse relayer if there's a match
+  // const relayer = runtime.Relayers.find((r: RelayerRunObj) => r.Name === 'ibc-relayer')
+  const relayer = await setupIbcRelayer(undefined, config.runtime, [singleHopPath], false)
```
Comment (nicopernas) [inline src/lib/vibc/channel.ts:80]: Do we need to call `setupIbcRelayer(undefined)` here? Can we init it up earlier in the process and  "reuse" it here like with the vibc relayer?

File: src/lib/ibc/channel.ts [hunk 6]
```diff
@@ -206,7 +150,7 @@ class IbcChannelHandshaker {
     const openinit = self.cosmos.client.polyibc.MsgOpenIBCChannelResponseSchema.parse(event)
     log.debug(`openinit: ${JSON.stringify(openinit, null, 2)}`)
     this.channelId = openinit.channel_id
-    this.connectionHops = openinit.connection_id.split(/[/\.]/)
+    this.connectionHops = openinit.connection_id.split(/[/.]/)
   }
 
   async channOpenTry(counter: IbcChannelHandshaker, connectionHops: string[], eventName: string) {
```
Comment (nicopernas) [inline src/lib/ibc/channel.ts:156]: We should also remove all this code since this is what the ibc-relayer will do now

File: src/lib/vibc/channel.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+  // step 2: start vibc relayer and trigger channel init
+  const counterpartyConnectionHops = [virtualConnectionId, relayer.pathConfigs[0].config.dst['connection-id']]
+  log.verbose(
+    `starting vibc relayer with connection hops: ${counterpartyConnectionHops[0]} -> ${counterpartyConnectionHops[1]}`
+  )
+  await vibcB.startRelaying(counterpartyConnectionHops)
+
+  // TODO: use the relayer for this?
+  await ibcA.channOpenInit(polymer, connectionHops)
```
Comment (nicopernas) [inline src/lib/vibc/channel.ts:152]: This should be done by the relayer indeed. If not, the sdk is not fully removed from the process and we don't have a way for it to work in the real world. Isn't there a `rly` command we can call into? if not, maybe we should add one.

File: src/tests/devnet/cli_e2e.spec.ts [hunk 8]
```diff
@@ -64,11 +64,12 @@ async function waitForEvent(t: any, chainName: string, minHeight: string, eventN
...
     },
     10,
-    6_000
+    6_000,
+    `failed to find event '${eventName}' on ${chainName}`
   )
 }
 
-test('cli end to end: eth <-> polymer <-> wasm', async (t) => {
+async function testEndToEnd(t: any, vIbcToIbc: boolean) {
```
Comment (nicopernas) [inline src/tests/devnet/cli_e2e.spec.ts:72]: What's the reason for this change? the whole point of doing it the way we were is that we can test how multiple channels are created within the same chains. This is defeating that purpose.

File: scripts/debug_relayer.sh [hunk 9]
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (7AC) [reply scripts/debug_relayer.sh:1]: The parts that are specific to the SDK are:
- it expects polymer and wasm to be in this order in the relayer config
- it expects them to be started via docker with bridge as a network so it extracts the ports accordingly

If it was more generic I'd agree it would belong to the relayer, in this case there are hardcoded bits based on sdk defaults so I could go either way I guess, one thing to keep in mind though is if we were to upstream relayer changes they'd likely not want this (nor it's really useful if you're operating from a relayer workspace alone).

File: src/cli/commands.ts [hunk 10]
```diff
@@ -256,6 +266,47 @@ type ChannelOpts = {
   chainB: { chainId: string; portID: string; version: string }
 }
 
+async function channelHandshake(runtime: ChainSetsRunObj, endpointA: Endpoint, endpointB: Endpoint) {
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/cli/commands.ts:269]: I put it here because it routes between `lib/vibc/channel.ts` and `lib/ibc/channel.ts` based on the type of channel. There's no `cli/channel.ts` anymore, feels like adding one just for this function is overkill?

File: src/lib/cosmos/cli_client.ts [hunk 11]
```diff
@@ -25,15 +25,32 @@ export class CosmosChainClient {
     return parsed.balances
   }
 
+  // TODO: this is a hack, we should be using grpc instead and return IdentifiedConnection
+  // https://github.com/open-ibc/ibc-sdk/issues/162
   async ibcConnections(): Promise<{
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/cosmos/cli_client.ts:30]: `vibcPathFromChainClients()` in `ibc/path.ts`

File: src/lib/docker.ts [hunk 12]
```diff
@@ -246,11 +246,14 @@ export class Container {
       const proc = $`${allArgs}`
       if (typeof stdincb !== 'undefined') stdincb(proc.stdin)
       const out = await proc
-      if (out.stdout) {
-        log.debug(`stdout: ${out.stdout}`)
+      if (out.stderr.length > 0) {
+        log.warn(`${out.stderr}`)
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/docker.ts:250]: Yeah you're right, was looking for a less verbose way to distinguish between the two. How about stdout -> verbose and stderr -> debug instead?

File: src/lib/docker.ts [hunk 13]
```diff
@@ -246,11 +246,14 @@ export class Container {
...
       }
-      if (out.stderr) {
-        log.debug(`stderr: ${out.stderr}`)
+      if (out.stdout.length > 0) {
+        log.debug(`${out.stdout}`)
+      }
+      if (out.exitCode !== 0) {
+        throw new Error(`failed to exec command: ${allArgs.map($.quote).join(' ')}`)
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/docker.ts:256]: Do we have such cases? The reason I did this is some errors weren't showing up at all. How about we leave it this way and revisit when we actually have benign non-zero return codes?

File: src/lib/ibc/channel.ts [hunk 14]
```diff
@@ -1,37 +1,21 @@
...
+import { EventsFilter, TxEvent } from '../query'
 import { TextEncoder } from 'util'
-import { flatCosmosEvent, getLogger, waitForBlocks } from '../lib/utils'
-import { addressify } from '../lib/ethers'
+import { flatCosmosEvent, getLogger, waitForBlocks } from '../utils'
 
 const log = getLogger()
 
-type Endpoint = {
+export type Endpoint = {
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/ibc/channel.ts:13]: ends up being `ibc.channel.Endpoint` which seems ok, any ideas?

File: src/lib/ibc/channel.ts [hunk 15]
```diff
@@ -1,37 +1,21 @@
-import {
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/ibc/channel.ts:1]: The problem was the file was really big and it was becoming hard to follow, and I would have added a lot of the new logic there too. I think smaller/more specialized packages/files/classes is generally better, also in this case having `ibc` and `vibc` split makes sense imho as one could build on top of the other making it more modular. Current state of affairs is:

```
src/lib/ibc:
channel.ts      path.ts

src/lib/vibc:
channel.ts
```

We could merge everything under a `vibc` package and have a single `channel.ts` but kind of feels like a step back to me. I came to this without knowing the code and feel like having smaller `channel.ts` files really helped me read it, the next person could have a similar experience.

File: src/lib/ibc/path.ts [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+  get name(): string {
+    return `${this.src.name}-${this.dst.name}`
+  }
+
+  reverse() {
+    return new Path(this.dst, this.src, this.hop.reverse())
+  }
+
+  toString(): string {
+    let output = `{"src": ${this.src.toString()}, "dst": ${this.dst.toString()}`
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/ibc/path.ts:66]: good idea, much cleaner

File: src/lib/relayers.ts [hunk 17]
```diff
@@ -56,21 +57,30 @@ async function setupIbcTsRelayer(runtime: ChainSetsRunObj, relayPath: Tuple) {
...
 }
 
-export async function setupIbcRelayer(runtime: ChainSetsRunObj, paths: Tuple[]) {
-  log.info(`setting up ibc-relayer with path(s) ${paths.map((p) => `${p[0]} -> ${p[1]}`).join(', ')}`)
+export async function setupIbcRelayer(
+  relayer: IbcRelayer | undefined,
+  runtime: ChainSetsRunObj,
+  paths: Path[],
+  start: boolean = true
+): Promise<IbcRelayer> {
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/relayers.ts:65]: Unfortunately no as they'll have different configs: the first time it doesn't need to start, just create the paths with the missing IBC connections, the second time it actually gets the full config (which unfortunately is composed from 2-3 different places so can't be done in one shot) and it needs to actually start relaying. I'm abusing `setupIbcRelayer()` here, conceptually what's happening during each of the two calls is really:
- create clients/connections for single hop path polymer-wasm (whichever direction)
- create multihop channel and start relaying with multihop config (path above + 2 more)

I think there's a lot of room for improvement/refactoring not only here, mostly around https://github.com/open-ibc/ibc-sdk/issues/58: not trying to reuse containers would simplify use cases such as this one. My 2c is we should defer all that for now.

File: src/lib/relayers/ibc.ts [hunk 18]
```diff
@@ -2,33 +2,89 @@ import path from 'path'
...
       workDir: '/tmp',
       entrypoint: 'sh',
       volumes: [[containerDir, '/tmp']]
-    })
+    }
+    let binary = 'rly'
+    if (process.env.IBC_RELAYER_BIN_DIR !== undefined) {
+      containerConfig.volumes.push([process.env.IBC_RELAYER_BIN_DIR, '/usr/local/bin'])
+      binary = '/usr/local/bin/rly'
+    }
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/relayers/ibc.ts:84]: It's not really working on macOS due to some arch differences that I didn't debug, but it was meant to inject a locally built binary to iterate faster. I can leave it and debug later (or maybe it already works on Linux) or remove, your call

File: src/lib/relayers/ibc.ts [hunk 19]
```diff
@@ -55,74 +111,144 @@ export class IBCRelayer {
...
-        dst: {
-          'chain-id': dst
-        },
-        'src-channel-filter': {
-          rule: null,
-          'channel-list': []
+        const account = chain.Accounts.find((a) => a.Name === 'relayer')
+        if (!account) {
+          log.error(`could not find relayer account on chain '${chain.Name}'`)
+          continue
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/relayers/ibc.ts:136]: good idea, let me throw an error actually

File: src/lib/relayers/ibc.ts [hunk 20]
```diff
@@ -55,74 +111,144 @@ export class IBCRelayer {
...
+        await this.exec([this.binary, 'keys', 'restore', chain.Name, 'default', account.Mnemonic!])
+      }
+    } else {
+      log.debug('cleaning up relay paths to reuse IBC relayer')
+      for (const config of this.pathConfigs) {
+        log.debug(`deleting path ${config.name}`)
+        await this.exec([this.binary, 'paths', 'delete', config.name])
+      }
+      this.pathConfigs = []
+    }
```
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/relayers/ibc.ts:147]: no this is within a single call to `ibctl channel`

File: src/lib/relayers/ibc.ts (hunk 1 above)
Comment (7AC) [reply src/lib/relayers/ibc.ts:183]: I did, it still works because there are no other tendermint clients when it runs. Let me fix it

File: src/lib/relayers/ibc.ts (hunk 2 above)
Comment (7AC) [reply src/lib/relayers/ibc.ts:179]: sounds good

File: src/lib/relayers/ibc.ts (hunk 3 above)
Comment (7AC) [reply src/lib/relayers/ibc.ts:190]: yeah let me check

Comment (7AC) [reply src/lib/relayers/ibc.ts:190]: mm actually prob not, we would need another waitFor() to wait for transactions to be executed. I'll enhance the comment for now

File: src/lib/relayers/ibc.ts (hunk 4 above)
Comment (7AC) [reply src/lib/relayers/ibc.ts:201]: what do you suggest? I used these so we could only touch `pathConfig` when we're sure we have all of them

File: src/lib/vibc/channel.ts (hunk 5 above)
Comment (7AC) [reply src/lib/vibc/channel.ts:80]: Not really as it's doing two different things per https://github.com/open-ibc/ibc-sdk/pull/129#discussion_r1285705657

File: src/lib/ibc/channel.ts (hunk 6 above)
Comment (7AC) [reply src/lib/ibc/channel.ts:156]: I left the ts-relayer in and it can be enabled via `$IBC_RELAYER`, we can delete all those when we clean that up.

File: src/lib/vibc/channel.ts (hunk 7 above)
Comment (7AC) [reply src/lib/vibc/channel.ts:152]: Yes there's a command to create a channel, in production we can use that. For here now that I think of it this is actually technically better as it's using `StargateClient` instead of shelling out to a command

File: src/tests/devnet/cli_e2e.spec.ts (hunk 8 above)
Comment (7AC) [reply src/tests/devnet/cli_e2e.spec.ts:72]: yeah doing multiple channels doesn't really work, will enable it in a separate PR and enhance both tests with it

File: scripts/debug_relayer.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (nicopernas) [reply scripts/debug_relayer.sh:1]: You added this sdk specific parts for simplicity since that's the env you are testing the relayer with. But the point of the script is to debug the relayer and that has no place in this repo. If anything, it should go into the relayer repo. If it can't be upstreamed, that's another story. In which case, you could keep it in your local workspace.

File: src/cli/commands.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (nicopernas) [reply src/cli/commands.ts:269]: I'd argue that there shouldn't be `lib/vibc/channel.ts` and `lib/ibc/channel.ts` in the first place (as pointed out in other comments)

File: src/lib/docker.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nicopernas) [reply src/lib/docker.ts:250]: There shouldn't be a difference between these two. Some containers will only log to stderr and others to stdout. If you wanna log these, pick verbose or debug level for both.

File: src/lib/docker.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (nicopernas) [reply src/lib/docker.ts:256]: In that case we should have fixed the callers instead :) 

First case that comes to mind is the `exec` command. Try running `grep` with it like so:
```
$ ibctl exec poly grep foo /etc/alpine-release
```

The situation is even worse than I thought though. That command will throw an error **without** your patch and it totally shouldn't
```
$ ibctl exec poly grep foo /etc/alpine-release
[11:45:01.420 error]: docker container exec 67c253274262 grep foo /etc/alpine-release
[11:45:01.420 error]:
    at Container.<anonymous> (/Users/nicolas/workspace/ibc-sdk/bin/ibctl:167627:38)
    exit code: 1
```

We shouldn't call into `zx` like we currently are. We should wrap the call with `zx.nothrow()` and handle the result gracefully. In which case your patch breaks in the same way

```diff
diff --git a/src/lib/docker.ts b/src/lib/docker.ts
index e09e374..5523280 100644
--- a/src/lib/docker.ts
+++ b/src/lib/docker.ts
@@ -243,7 +244,7 @@ export class Container {
     allArgs.push(...cmds)
     log.debug(allArgs.map($.quote).join(' '))
     try {
-      const proc = $`${allArgs}`
+      const proc = zx.nothrow($`${allArgs}`)
       if (typeof stdincb !== 'undefined') stdincb(proc.stdin)
       const out = await proc
       if (out.stdout) {
@@ -252,6 +253,9 @@ export class Container {
       if (out.stderr) {
         log.debug(`stderr: ${out.stderr}`)
       }
+      if (out.exitCode !== 0) {
+        throw new Error(`failed to exec command: ${allArgs.map($.quote).join(' ')}`)
+      }
       return out
     } catch (e) {
       log.error(allArgs.map($.quote).join(' '))
```
```sh
$ ibctl exec poly grep foo /etc/alpine-release
[11:57:56.617 error]: docker container exec 67c253274262 grep foo /etc/alpine-release
[11:57:56.617 error]: failed to exec command: docker container exec 67c253274262 grep foo /etc/alpine-release
```

If you want to check for the error code still, we should allow the caller to somehow decide whether `out.exitCode !== 0` means error or not.

File: src/lib/relayers.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (nicopernas) [reply src/lib/relayers.ts:65]: You don't need to create multiple containers to do so as suggested in #58, you simply call the command you need on the running container as many times as needed and only call `start` once you are ready. 
I'm asking because I upgraded the vibc-relayer cli following the exact same commands `rly` uses. Granted they use different configs, but that shouldn't matter.

Note that the sdk is effectively calling into the vibc-relayer container many times, from different places. 

```
vibc-relayer chains add ...
vibc-relayer config set ... 
vibc-relayer paths add ...
vibc-relayer paths update ... 
...
vibc-relayer start
```

File: scripts/debug_relayer.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (7AC) [reply scripts/debug_relayer.sh:1]: > You added this sdk specific parts for simplicity since that's the env you are testing the relayer with. But the point of the script is to debug the relayer and that has no place in this repo

I'm not sure I agree with this. The relayer is the default relayer in the sdk, the script is roughly 10 lines out of which 9 are sdk-specific. I found it very useful to debug, also people could reuse the logic for the ports for other components, if I keep it in my workspace I worry people will spend time rewriting some of the same logic in the future. Is there a way I can change it that would make you more comfortable about putting it in the repo?

File: src/cli/commands.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (7AC) [reply src/cli/commands.ts:269]: I can merge the two, tho my personal experience was that `channel.ts` was too long to be manageable. What's the rationale for not separating regular ibc vs vibc?

File: src/lib/docker.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/docker.ts:250]: Aren't there containers that log to both? If they do there's value in looking at stderr (more critical usually) and maybe ignoring stdout in some cases. If there are no containers that lot to both I agree we don't need to distinguish.

File: src/lib/docker.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (7AC) [reply src/lib/docker.ts:256]: I would go one step further and do https://github.com/open-ibc/ibc-sdk/issues/61, then it's entirely up to the caller what to do with the result. In this intermediate state I'd argue seeing the errors is better than not seeing them (making debug harder), but I hear your concern about not making it confusing for the user. I'll revert this change.

**Review by nicopernas — APPROVED** [review]
